source/declarations.tex

```diff
@@ -2868,159 +2868,132 @@
-contains the \tcode{auto} \grammarterm{type-specifier},
-the program is ill-formed.
-\tcode{T}
-is called the array
-\term{element type};
-this type shall not be a reference type, \cv{}~\tcode{void},
-or a function type.
-\indextext{declaration!array}%
-If the
-\grammarterm{constant-expression}\iref{expr.const}
-is present, it shall be a converted constant
-expression of type \tcode{std::size_t} and
-its value shall be greater than zero.
-The constant expression specifies the
-\indextext{array!bound}%
+and the type of the contained \grammarterm{declarator-id}
+in the declaration \tcode{T} \tcode{D1}
+is ``\placeholder{derived-declarator-type-list} \tcode{T}'',
+the type of the \grammarterm{declarator-id} in \tcode{D} is
+``\placeholder{derived-declarator-type-list} array of \tcode{N} \tcode{T}''.
+The \grammarterm{constant-expression}
+shall be a converted constant expression of type \tcode{std::size_t}\iref{expr.const}.
 \indextext{bound, of array}%
-\term{bound}
-of (number of elements in) the array.
-If the value of the constant expression is
-\tcode{N},
-the array has
-\tcode{N}
-elements numbered
-\tcode{0}
-to
-\tcode{N-1},
-and the type of the identifier of
-\tcode{D}
-is ``\placeholder{derived-declarator-type-list} array of
-\tcode{N}
-\tcode{T}''.
-An object of array type contains a contiguously allocated non-empty set of
-\tcode{N}
-subobjects of type
-\tcode{T}.
-Except as noted below, if
-the constant expression is omitted, the type of the identifier of
-\tcode{D}
-is ``\placeholder{derived-declarator-type-list} array of unknown bound of
-\tcode{T}'',
-an incomplete object type.
-The type ``\placeholder{derived-declarator-type-list} array of
-\tcode{N}
-\tcode{T}''
-is a different type from the type
-``\placeholder{derived-declarator-type-list} array of unknown bound of
-\tcode{T}'',
-see~\ref{basic.types}.
+Its value \tcode{N} specifies the \defnx{array bound}{array!bound},
+i.e., the number of elements in the array;
+\tcode{N} shall be greater than zero.
+An object of such a type contains
+a contiguously allocated non-empty set
+of \tcode{N} subobjects of type \tcode{T},
+known as the \defnx{elements}{array!element} of the array,
+and numbered \tcode{0} to \tcode{N-1}.
+
+\pnum
+In a declaration \tcode{T} \tcode{D} where \tcode{D} has the form
+
+\begin{ncsimplebnf}
+\terminal{D1 [ ]} \opt{attribute-specifier-seq}
+\end{ncsimplebnf}
+
+and the type of the contained \grammarterm{declarator-id}
+in the declaration \tcode{T} \tcode{D1}
+is ``\placeholder{derived-declarator-type-list} \tcode{T}'',
+the type of the \grammarterm{declarator-id} in \tcode{D} is
+``\placeholder{derived-declarator-type-list} array of unknown bound of \tcode{T}'', except as specified below.
+
+\pnum
+\indextext{declaration!array}%
+In either case, the type is an \defn{array type}.
+The optional \grammarterm{attribute-specifier-seq}
+appertains to the array type.
+
+\pnum
+\tcode{T} is called the array \defn{element type};
```

Comment:
This is also wrong; `T` isn't the element type. Given `int *p[3];`, the element type is `int*`, but the only element type computed by this algorithm for that declaration is `int`.

Reply:
That does not seem to be accurate. T=`int *`; this was already computed by [dcl.ptr] before we even got here. Note that the initial parsing stuff gets us "array of N T" (or "array of unknown bound of T", so assuming this is a correct decomposition (I believe it is), T is actually `int *`.

Reply:
Oh right, I see it now. This is confusing because the `T` here is a different `T` from that used in paragraph 1 and 2 (and indeed throughout the rest of [dcl.meaning]). In those other places, `T` is the type determined from the *type-specifier*s in the *decl-specifier-seq*, but here it's the eventual type following the ''array of N'' -- that is, it's the `T` introduced in p3, which is just shadowing the `T` from p1 and p2 and is in general a different type.

I think this would be much clearer if you replaced `T` with (say) `U` in p3 onwards.

Reply:
Sorry, I still don't get it. In p1 and p2, we compute "(mumble) array of N T", so the T from p1 and p2 is actually the array element type. I do agree that p1 and p2 may not actually describe an array type, but something like a pointer to an array type, but that doesn't change the fact that T is the element type of the array part of the type.
Consider p1: "derived-declarator-type-list array of N T”. 
and p2: "derived-declarator-type-list array of unknown bound of T".

So, re-use of the T seems to be totally fine to me.

Reply:
I think the problem is that p1 and p2 only computes the type in the declaration of the form `T D1[n]`.

For `int *p[3]`, the procedure works by first computing the type in `int p[3]`, then adjusting the type by replacing final "int" with "pointer to int". The latter is specified by the first sentence in [dcl.ptr]/1.

Reply:
Ah, I think I see it now.  p1+p2 would never see "T=pointer to int" for your example, but the statements in p3+ are valid for any kind of array, not just for the T's we've discovered in p1+p2.

Reply:
Renamed T to U in p3+.